include/ofi_uffd.h

```diff
@@ -0,0 +1,209 @@
+/*
+ * Copyright (c) 2019 Amazon.com, Inc. or its affiliates. All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#ifndef _OFI_UFFD_H_
+#define _OFI_UFFD_H_
+
+#if HAVE_LINUX_USERFAULTFD_H
+
+#include <config.h>
+#include <stdlib.h>
+#include <ofi_util.h>
+#include <errno.h>
+#include <sys/syscall.h>
+#include <sys/ioctl.h>
+#include <linux/userfaultfd.h>
+
+#define MB(x)((size_t)(x) << 20)
+#define GB(x)((size_t)(x) << 30)
+
+/* See TODO in function ofi_uffd_register */
+static long page_sizes[] = {4096, MB(2), GB(1)};
+
+static long get_page_size(void)
+{
+	long ret;
+
+	/* sysconf can return -1 and not change errno */
+	errno = 0;
+	ret = sysconf(_SC_PAGESIZE);
+
+	if (ret <= 0) {
+		FI_WARN(&core_prov, FI_LOG_MR,
+			"sysconf/pagesize: ret %ld, err %s\n",
+			ret, strerror(errno));
+		if (errno)
+			return -errno;
+		else
+			return -FI_EOTHER;
+	}
+
+	return ret;
+}
+
+static inline void set_uffdio_range(struct uffdio_range *range,
+				    uint64_t start, uint64_t len,
+				    long page_size)
+{
+	range->start = (start / page_size) * page_size;
+	range->len = ofi_div_ceil(len, page_size) * page_size;
+}
+
+/*
+ * TODO: ioctl register needs to be passed page aligned boundaries.
+ * Need to find a more sane way of detecting a region's page size.
+ */
+static inline int ofi_uffd_register(int uffd, uint64_t start, uint64_t len)
```

Comment:
From the perspective of the rest of the code, I would anticipate start as void* and len as size_t.

Reply:
The range struct looks like this
```
struct uffdio_range {
	__u64 start;
	__u64 len;
};
```

I wanted to avoid casting it inside this file.

Reply:
Why?  The cast is needed somewhere.  I would have expected this code to abstract away the userfaultfd details from the rest of the infrastructure, but this is exposing those details out.